File: server/aiService.ts
/**
 * AIService — Ported from com.icare.service.AIService (Java)
 *
 * Keyword-based symptom→specialty mapping and offline hospital triage.
 * Uses the real hospital data from ExcelService.
 */

import { Hospital, Doctor, getHospitals } from "./dbService.js";
import fs from "fs";
import path from "path";

// ─── Types ───────────────────────────────────────────────────────

interface TriageCase {
  id: string;
  complaint: string;
  branches: Array<{
    id: string;
    condition: string;
    severity: number[];
    triage: number;
    spec: string;
    action: string;
  }>;
}

interface TriageData {
  cases: TriageCase[];
  triage_levels: Array<{ code: number; label: string; color: string }>;
  red_flags: Array<{ id: string; text: string; escalate: number }>;
}

interface SpecialtiesData {
  specs: Record<string, string>;
}

// ─── Symptom → Specialty Rules ───────────────────────────────────

const SYMPTOM_SPECIALTY_RULES: [string, string][] = [
  ["chest pain", "Cardiology"],
  ["heart", "Cardiology"],
  ["shortness of breath", "Pulmonology"],
  ["difficulty breathing", "Pulmonology"],
  ["breath", "Pulmonology"],
  ["fever", "Internal Medicine"],
  ["cough", "Internal Medicine"],
  ["infection", "Internal Medicine"],
  ["headache", "Neurology"],
  ["migraine", "Neurology"],
  ["stomach", "Gastroenterology"],
  ["abdominal", "Gastroenterology"],
  ["nausea", "Gastroenterology"],
  ["dizzy", "General Practice"],
  ["other", "General Practice"],
];

// ─── Specialty Keywords ──────────────────────────────────────────

const SPECIALTY_KEYWORDS: Record<string, string[]> = {
  cardiology: ["cardiology", "cardiac", "heart", "cardio"],
  pulmonology: ["pulmonology", "pulmonary", "respiratory", "lung", "lungs", "breathing"],
  "internal medicine": ["internal medicine", "internal", "general medicine", "medicine", "general internal"],
  neurology: ["neurology", "neuro", "brain", "headache", "migraine", "nerve", "stroke"],
  gastroenterology: ["gastroenterology", "gastro", "digestive", "stomach", "abdominal", "nausea", "gut", "bowel"],
  "general practice": ["general practice", "general", "family", "primary", "primary care", "gp", "family medicine", "general medicine"],
};

// ─── JSON Data Loading ───────────────────────────────────────────

let triageData: TriageData | null = null;
let specialtiesData: SpecialtiesData | null = null;

function loadJsonFile<T>(filename: string): T | null {
  const candidates = [
    path.resolve(process.cwd(), "src", "data", filename),
    path.resolve(process.cwd(), filename),
    path.resolve(process.cwd(), "..", filename),
  ];

  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      const content = fs.readFileSync(candidate, "utf-8");
      return JSON.parse(content) as T;
    }
  }

  console.warn(`[AIService] Could not find ${filename}`);
  return null;
}

function ensureDataLoaded(): void {
  if (!triageData) {
    triageData = loadJsonFile<TriageData>("triage.json");
  }
  if (!specialtiesData) {
    specialtiesData = loadJsonFile<SpecialtiesData>("specialties.json");
  }
}

// ─── Helpers ─────────────────────────────────────────────────────

function getKeywordsForSpecialty(specialty: string): string[] {
  if (!specialty) return [];
  return SPECIALTY_KEYWORDS[specialty.toLowerCase()] || [specialty.toLowerCase()];
}

function matchesAnyKeyword(target: string | undefined, keywords: string[]): boolean {
  if (!target?.trim()) return false;
  const normalized = target.toLowerCase();
  return keywords.some(kw => kw && normalized.includes(kw));
}

function mapSymptomToSpecialty(symptoms: string): string {
  const normalized = (symptoms || "").toLowerCase();
  if (!normalized) return "General Practice";

  for (const [keyword, specialty] of SYMPTOM_SPECIALTY_RULES) {
    if (normalized.includes(keyword)) {
      return specialty;
    }
  }
  return "General Practice";
}

function matchesHospitalSpecialty(hospital: Hospital, lowerSpecialty: string, _lowerSymptoms: string): boolean {
  const keywords = getKeywordsForSpecialty(lowerSpecialty);

  // Check hospital services
  if (hospital.services?.some(service => matchesAnyKeyword(service, keywords))) {
    return true;
  }

  // Check hospital doctors
  if (hospital.doctors) {
    for (const doctor of hospital.doctors) {
      if (!doctor) continue;
      if (matchesAnyKeyword(doctor.department, keywords)) return true;
      if (matchesAnyKeyword(doctor.type, keywords)) return true;
      if (doctor.tags?.some(tag => matchesAnyKeyword(tag, keywords))) return true;
    }
  }

  return false;
}

function findHospitalsForSpecialty(specialty: string, symptoms: string): Hospital[] {
  const lowerSpecialty = specialty.toLowerCase();
  const lowerSymptoms = (symptoms || "").toLowerCase();
  return getHospitals().filter(h => matchesHospitalSpecialty(h, lowerSpecialty, lowerSymptoms));
}

function buildHospitalSummary(hospital: Hospital): Record<string, unknown> {
  return {
    id: hospital.id,
    name: hospital.name,
    address: hospital.address,
    type: hospital.level?.trim() || "Hospital",
    lat: hospital.latitude,
    lng: hospital.longitude,
    specializations: hospital.services || [],
    contactNumber: hospital.contactNumber,
    doctors: hospital.doctors
      .map(d => d.name)
      .filter(Boolean)
      .slice(0, 3),
  };
}

// ─── Public API ──────────────────────────────────────────────────

export function getTriageResponse(userInput: string): string {
  ensureDataLoaded();

  if (triageData?.cases) {
    for (const triageCase of triageData.cases) {
      const complaint = triageCase.complaint.toLowerCase();
      if (userInput.toLowerCase().includes(complaint)) {
        const branch = triageCase.branches[0];
        return (
          `Based on your symptoms, this seems ${branch.condition}. ` +
          `Recommended action: ${branch.action}. ` +
          `Specialty: ${getSpecialtyName(branch.spec)}`
        );
      }
    }
  }

  return "Please provide more details about your symptoms.";
}

export function getSpecialtyName(specId: string): string {
  ensureDataLoaded();
  if (specialtiesData?.specs && specId in specialtiesData.specs) {
    return specialtiesData.specs[specId];
  }
  return "General Practitioner";
}

export function offlineTriage(symptoms: string): Record<string, unknown> {
  const cleaned = (symptoms || "").trim();
  const specialty = mapSymptomToSpecialty(cleaned);
  const matchedHospitals = findHospitalsForSpecialty(specialty, cleaned);

  if (matchedHospitals.length > 0) {
    return {
      reply: "These are the suited hospitals for that section. All recommendations are based on the loaded hospital database.",
      specialty,
      location: "Hospital database region",
      hospitals: matchedHospitals.map(buildHospitalSummary),
      found: true,
    };
  }

  return {
    reply: "The specific specialist or hospital for this condition may not be available in the current hospital database.",
    hospitals: [],
    found: false,
  };
}

File: backend/src/main/java/com/icare/service/AIService.java
package com.icare.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.icare.model.Doctor;
import com.icare.model.Hospital;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class AIService {

    private final ObjectMapper objectMapper = new ObjectMapper();
    private JsonNode specialties;
    private JsonNode triageData;
    private final ExcelService excelService;

    private static final LinkedHashMap<String, String> SYMPTOM_SPECIALTY_RULES = createSymptomSpecialtyRules();
    private static final Map<String, List<String>> SPECIALTY_KEYWORDS = createSpecialtyKeywords();

    @Autowired
    public AIService(ExcelService excelService) {
        this.excelService = excelService;
        try {
            specialties = loadJsonResource("ph_medical_specialties_mvp.json");
            triageData = loadJsonResource("ph_nurse_ai_dataset_mvp.json");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private JsonNode loadJsonResource(String filename) throws IOException {
        Path path = Paths.get(filename);
        if (!Files.exists(path)) {
            path = Paths.get("..", filename);
        }
        if (Files.exists(path)) {
            return objectMapper.readTree(Files.readAllBytes(path));
        }

        ClassPathResource resource = new ClassPathResource(filename);
        if (resource.exists()) {
            try (InputStream stream = resource.getInputStream()) {
                return objectMapper.readTree(stream);
            }
        }

        throw new IOException("Missing resource: " + filename + " in current folder, parent folder, or classpath.");
    }

    private static LinkedHashMap<String, String> createSymptomSpecialtyRules() {
        LinkedHashMap<String, String> rules = new LinkedHashMap<>();
        rules.put("chest pain", "Cardiology");
        rules.put("heart", "Cardiology");
        rules.put("shortness of breath", "Pulmonology");
        rules.put("difficulty breathing", "Pulmonology");
        rules.put("breath", "Pulmonology");
        rules.put("fever", "Internal Medicine");
        rules.put("cough", "Internal Medicine");
        rules.put("infection", "Internal Medicine");
        rules.put("headache", "Neurology");
        rules.put("migraine", "Neurology");
        rules.put("stomach", "Gastroenterology");
        rules.put("abdominal", "Gastroenterology");
        rules.put("nausea", "Gastroenterology");
        rules.put("dizzy", "General Practice");
        rules.put("other", "General Practice");
        return rules;
    }

    private static Map<String, List<String>> createSpecialtyKeywords() {
        Map<String, List<String>> keywords = new HashMap<>();
        keywords.put("cardiology", List.of("cardiology", "cardiac", "heart", "cardio"));
        keywords.put("pulmonology", List.of("pulmonology", "pulmonary", "respiratory", "lung", "lungs", "breathing"));
        keywords.put("internal medicine", List.of("internal medicine", "internal", "general medicine", "medicine", "general internal"));
        keywords.put("neurology", List.of("neurology", "neuro", "brain", "headache", "migraine", "nerve", "stroke"));
        keywords.put("gastroenterology", List.of("gastroenterology", "gastro", "digestive", "stomach", "abdominal", "nausea", "gut", "bowel"));
        keywords.put("general practice", List.of("general practice", "general", "family", "primary", "primary care", "gp", "family medicine", "general medicine"));
        return keywords;
    }

    private List<String> getKeywordsForSpecialty(String specialty) {
        if (specialty == null) {
            return Collections.emptyList();
        }
        return SPECIALTY_KEYWORDS.getOrDefault(specialty.toLowerCase(Locale.ROOT), Collections.singletonList(specialty.toLowerCase(Locale.ROOT)));
    }

    private boolean matchesAnyKeyword(String target, List<String> keywords) {
        if (target == null || target.isBlank()) {
            return false;
        }
        String normalized = target.toLowerCase(Locale.ROOT);
        for (String keyword : keywords) {
            if (keyword != null && !keyword.isBlank() && normalized.contains(keyword)) {
                return true;
            }
        }
        return false;
    }

    // Simple keyword matching for RAG
    public String getTriageResponse(String userInput) {
        // Match to cases based on keywords
        for (JsonNode caseNode : triageData.get("cases")) {
            String complaint = caseNode.get("complaint").asText().toLowerCase();
            if (userInput.toLowerCase().contains(complaint)) {
                // Return first branch for simplicity
                JsonNode branch = caseNode.get("branches").get(0);
                return "Based on your symptoms, this seems " + branch.get("condition").asText() +
                       ". Recommended action: " + branch.get("action").asText() +
                       ". Specialty: " + getSpecialtyName(branch.get("spec").asText());
            }
        }
        return "Please provide more details about your symptoms.";
    }

    public String getSpecialtyName(String specId) {
        JsonNode specs = specialties.get("specs");
        return specs.has(specId) ? specs.get(specId).asText() : "General Practitioner";
    }

    // Lazy load specialty description
    public String getSpecialtyDescription(String specId) {
        // For MVP, return name only, or fetch from full JSON if needed
        return getSpecialtyName(specId);
    }

    // Call Gemini 1.5 API with minimal context
    public String callGemini(String prompt) {
        // Placeholder for Gemini API call
        // Use RestTemplate to call Google AI Studio API
        // For now, return mock response
        return "AI Response: " + prompt.substring(0, 50) + "...";
    }

    // Offline mode: checklist triage
    public Map<String, Object> offlineTriage(String symptoms) {
        String cleaned = symptoms == null ? "" : symptoms.trim();
        String specialty = mapSymptomToSpecialty(cleaned);
        List<Hospital> matchedHospitals = findHospitalsForSpecialty(specialty, cleaned);

        Map<String, Object> result = new HashMap<>();
        if (!matchedHospitals.isEmpty()) {
            result.put("reply", "These are the suited hospitals for that section. All recommendations are based on the loaded hospital database.");
            result.put("specialty", specialty);
            result.put("location", "Hospital database region");
            result.put("hospitals", matchedHospitals.stream()
                    .map(this::buildHospitalSummary)
                    .collect(Collectors.toList()));
            result.put("found", true);
        } else {
            result.put("reply", "The specific specialist or hospital for this condition may not be available in the current hospital database.");
            result.put("hospitals", Collections.emptyList());
            result.put("found", false);
        }
        return result;
    }

    private String mapSymptomToSpecialty(String symptoms) {
        String normalized = symptoms == null ? "" : symptoms.toLowerCase(Locale.ROOT);
        if (normalized.isBlank()) {
            return "General Practice";
        }
        for (Map.Entry<String, String> entry : SYMPTOM_SPECIALTY_RULES.entrySet()) {
            if (normalized.contains(entry.getKey())) {
                return entry.getValue();
            }
        }
        return "General Practice";
    }

    private List<Hospital> findHospitalsForSpecialty(String specialty, String symptoms) {
        String lowerSpecialty = specialty.toLowerCase(Locale.ROOT);
        String lowerSymptoms = symptoms == null ? "" : symptoms.toLowerCase(Locale.ROOT);
        return excelService.getHospitals().stream()
                .filter(hospital -> matchesHospitalSpecialty(hospital, lowerSpecialty, lowerSymptoms))
                .collect(Collectors.toList());
    }

    private boolean matchesHospitalSpecialty(Hospital hospital, String lowerSpecialty, String lowerSymptoms) {
        if (hospital == null) {
            return false;
        }

        List<String> keywords = getKeywordsForSpecialty(lowerSpecialty);
        boolean hasServiceMatch = hospital.getServices() != null && hospital.getServices().stream()
                .anyMatch(service -> matchesAnyKeyword(service, keywords));
        if (hasServiceMatch) {
            return true;
        }

        if (hospital.getDoctors() != null) {
            for (Doctor doctor : hospital.getDoctors()) {
                if (doctor == null) continue;
                if (matchesAnyKeyword(doctor.getDepartment(), keywords)) {
                    return true;
                }
                if (matchesAnyKeyword(doctor.getType(), keywords)) {
                    return true;
                }
                if (doctor.getTags() != null) {
                    for (String tag : doctor.getTags()) {
                        if (matchesAnyKeyword(tag, keywords)) {
                            return true;
                        }
                    }
                }
            }
        }

        return false;
    }

    private Map<String, Object> buildHospitalSummary(Hospital hospital) {
        Map<String, Object> summary = new HashMap<>();
        summary.put("id", hospital.getId());
        summary.put("name", hospital.getName());
        summary.put("address", hospital.getAddress());
        summary.put("type", hospital.getLevel() != null && !hospital.getLevel().isBlank() ? hospital.getLevel() : "Hospital");
        summary.put("lat", hospital.getLatitude());
        summary.put("lng", hospital.getLongitude());
        summary.put("specializations", hospital.getServices() != null ? hospital.getServices() : Collections.emptyList());
        summary.put("contactNumber", hospital.getContactNumber());
        summary.put("doctors", hospital.getDoctors().stream()
                .map(Doctor::getName)
                .filter(Objects::nonNull)
                .limit(3)
                .collect(Collectors.toList()));
        return summary;
    }
}